### PR TITLE
Remove casts in customized CG method

### PR DESCRIPTION
**Pull request by junchaoxia**
The PR tries to close the issue #121 by removing the casts in the customized CG method.

### CODE REVIEW

File: src/aspire/utils/optimize.py [hunk 1]
```diff
@@ -108,17 +108,17 @@ def identity(input_x):
     if np.any(x != 0):
         if cg_opt['verbose']:
             logger.info('[CG] Calculating initial residual')
-        a_x = a_fun(x)
+        a_x = np.array(a_fun(x), dtype=b.dtype)
```
Comment (janden) [inline src/aspire/utils/optimize.py]: Why is this cast necessary? Shouldn't `a_fun` respect the dtype of the input? If we do need to cast, it would be more natural to call `astype(b.dtype)`.

File: src/aspire/utils/optimize.py [hunk 2]
```diff
@@ -108,17 +108,17 @@ def identity(input_x):
...
     else:
-        a_x = np.zeros(x.shape)
+        a_x = np.zeros(x.shape, dtype=b.dtype)
 
     obj = (np.real(np.sum(x.conj() * a_x, -1)
             - 2 * np.real(np.sum(np.conj(b * x), -1))))
 
     if init['p'] is None:
-        p = s.copy()
+        p = np.array(s.copy(), dtype=b.dtype)
```
Comment (janden) [inline src/aspire/utils/optimize.py]: Again, why do we need to do this?

File: src/aspire/utils/optimize.py [hunk 3]
```diff
@@ -138,16 +138,15 @@ def identity(input_x):
         if cg_opt['verbose']:
             logger.info('[CG] Applying matrix & preconditioner')
 
-        a_p = a_fun(p)
-        old_gamma = np.real(np.sum(s.conj() * r, -1))
-
-        alpha = old_gamma / np.real(np.sum(p.conj() * a_p, -1))
+        a_p = np.array(a_fun(p), dtype=b.dtype)
+        old_gamma = np.sum(s.conj() * r, -1)
```
Comment (janden) [inline src/aspire/utils/optimize.py]: Why is the cast to real gone here? Same for `alpha` and `new_gamma`.

File: tests/test_optimize.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+
+def identity(x):
+    return x
+
+
+def precond_fun(S, x):
+    p = np.size(S, 0)
+    ensure(np.size(x) == p*p, 'The sizes of S and x are not consistent.')
+    x = m_reshape(x, (p, p))
```
Comment (janden) [inline tests/test_optimize.py]: Why are we using `m_reshape` here? From what I understand, these functions were introduced to help transition legacy code and shouldn't be used in new code.

File: tests/test_optimize.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+from aspire.utils import ensure
+from aspire.utils.matlab_compat import m_reshape
+from aspire.utils.optimize import conj_grad
+
+
+def identity(x):
+    return x
+
+
+def precond_fun(S, x):
```
Comment (janden) [inline tests/test_optimize.py]: Where do these complicated `precond_fun` and `apply` come from? Why not just do simple matrix multiplication? It's much easier to verify what's going on there.

File: tests/test_optimize.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        self.cg_opt = {
+            'shrinker': 'frobenius_norm', 'verbose': 0, 'max_iter': 250,
+            'iter_callback': [], 'store_iterates': False,
+            'rel_tolerance': 1e-12, 'precision': 'float64',
+            'preconditioner': identity
+        }
+
+        self.cg_opt["preconditioner"] = lambda x: precond_fun(self.S, x)
+
+        self.A = [
```
Comment (janden) [inline tests/test_optimize.py]: Generate these arrays and matrices using random numbers. No need to hardcode them.

File: tests/test_optimize.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
+import numpy as np
+from unittest import TestCase
+
+from aspire.utils import ensure
+from aspire.utils.matlab_compat import m_reshape
+from aspire.utils.optimize import conj_grad
+
+
+def identity(x):
```
Comment (janden) [inline tests/test_optimize.py]: Why is this needed? It's not used.

File: tests/test_optimize.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+    def testSolution(self):
+        self.assertTrue(np.allclose(apply(self.A, self.x), self.b))
+
+    def testConjGradReal(self):
+        covar_coeff, _, _ = conj_grad(lambda x: apply(self.A, x), self.b, self.cg_opt)
+
+        self.assertTrue(np.allclose(covar_coeff, self.x))
+
+    def testConjGradComplex(self):
```
Comment (janden) [inline tests/test_optimize.py:23]: We need to be testing not only for a complex right-hand side, but for a complex (positive semidefinite) matrix A.

**Review by janden — CHANGES_REQUESTED** [review]

File: src/aspire/utils/optimize.py (hunk 1 above)
Comment (junchaoxia) [reply src/aspire/utils/optimize.py]: The function crashed before I went through all dtype explicitly. I went back and removed some of them. The actual matter one is line 115,  `a_x = np.zeros(x.shape, dtype=b.dtype)`, for initializing `a_x`. Removed and used implicit passing in new version.

File: src/aspire/utils/optimize.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/utils/optimize.py]: Casting was removed in new version.

File: src/aspire/utils/optimize.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/utils/optimize.py]: This is the original problem discussed for casting. I will put back the casting for alpha and gamma and make them consist with Matlab code. By the way, there have been no casts for Line 144 an 145 for  `x` and `a_x`which are consistent with Matlab version.

File: tests/test_optimize.py (hunk 4 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: Removed.

File: tests/test_optimize.py (hunk 5 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: Previously there is no any unit test for the CG method. I think it is good to include them with preconditioner function.  I tried to use the Cov2D case to build the tests. The data and functions are reassembly the Cov2D problem.

File: tests/test_optimize.py (hunk 7 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: Removed in new version.

File: src/aspire/utils/optimize.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/utils/optimize.py]: It seems that casts in alpha and gamma need to be removed for complex matrix A as updates in new commit 6f43523.

File: tests/test_optimize.py (hunk 5 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: The `precond_fun` was removed in the new commit 6f43523.

File: tests/test_optimize.py (hunk 6 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: Done in new commit 6f43523.

File: tests/test_optimize.py (hunk 7 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: The `preconditioner` needs to be initialized for this customized CG method. the `precond_fun` was removed and the `identity` was restored in the commit 6f43523.

File: tests/test_optimize.py (hunk 8 above)
Comment (junchaoxia) [reply tests/test_optimize.py:23]: Complex matrix `A_comp` was added in new commit  6f43523.

File: src/aspire/utils/optimize.py (hunk 3 above)
Comment (janden) [reply src/aspire/utils/optimize.py]: Why are they needed? The Matlab version works fine with complex matrix.

File: tests/test_optimize.py (hunk 5 above)
Comment (janden) [reply tests/test_optimize.py]: Ok. I didn't say we should not test the preconditioner. I said let's not have a test case that's more complicated than necessary. A simple matrix multiplication problem *with preconditioner* will do.

File: tests/test_optimize.py (hunk 6 above)
Comment (janden) [reply tests/test_optimize.py]: What about `b`, `x`, and `x_comp`?

File: tests/test_optimize.py (hunk 7 above)
Comment (janden) [reply tests/test_optimize.py]: > The preconditioner needs to be initialized for this customized CG method.

That's a bug. Why isn't it provided as a default in `conj_grad`? Let's make an issue.

File: tests/test_optimize.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+        pass
+
+    def testSolution(self):
+        self.assertTrue(np.allclose(apply(self.A, self.x), self.b)
+                        and np.allclose(apply(self.A_comp, self.x_comp), self.b*1j))
+
+    def testConjGradReal(self):
+        covar_coeff, _, _ = conj_grad(lambda x: apply(self.A, x), self.b, self.cg_opt)
+
+        self.assertTrue(np.allclose(covar_coeff, self.x, atol=1e-6))
```
Comment (janden) [inline tests/test_optimize.py]: Why is this necessary?

**Comment by janden** [comment]
By the way, looks like #121 was largely fixed by 037bab239fff7975ee559d68e9936e731b35ff88.

File: src/aspire/utils/optimize.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/utils/optimize.py]: Restored casts for alpha and gamma in commit 82cff7a to be consistent with Matlab version.

File: tests/test_optimize.py (hunk 5 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: New unit tests were included in commit 6d8aeeb.

File: tests/test_optimize.py (hunk 6 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: All matrices (A and A_comp) and vectors (x and x_comp) were generated from random numbers in commit 6d8aeeb.  b and b_comp were evaluated from apply(A, x) and served as inputs to estimate x and x_comp.

Comment (vineetbansal) [reply tests/test_optimize.py]: We just talked .. @junchaoxia will make A/b/b_comp completely random (with no fixed seed) with the correct dtype/shape, and only check the solution. Also the `apply` function will go away.

Comment (junchaoxia) [reply tests/test_optimize.py]: In new commit 4947d02, all A and b are generated in different values for different unit tests.

File: tests/test_optimize.py (hunk 7 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: The Matlab version also needs  `identity` as default preconditioner. New issue #249 was created.

File: tests/test_optimize.py (hunk 9 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: This was used to pass the old unit tests for Mac OS. It is  not necessary for new unit tests.

File: tests/test_optimize.py (hunk 5 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: The unit tests were simplified to use the default CG option with the `identity` preconditioner  in commit 18bbf57.

File: tests/test_optimize.py (hunk 7 above)
Comment (junchaoxia) [reply tests/test_optimize.py]: The issue #249 was closed since we are using default `identity` preconditioner.

**Review by janden — DISMISSED** [review]
Looks good. I don't think there should be much conflict with #231, but let's discuss on Wednesday before merging.

**Comment by janden** [comment]
In the meantime, this also needs to be squashed since there's been a lot of changes back and forth that result in very little change.

**Comment by junchaoxia** [comment]
All the commits were rebased and squashed to one final commit  0ced3dc. Will perform merging after meeting on Wednesday.
